source/common/buffer/watermark_buffer.h

```diff
@@ -16,8 +16,10 @@ namespace Buffer {
 class WatermarkBuffer : public OwnedImpl {
 public:
   WatermarkBuffer(std::function<void()> below_low_watermark,
-                  std::function<void()> above_high_watermark)
-      : below_low_watermark_(below_low_watermark), above_high_watermark_(above_high_watermark) {}
+                  std::function<void()> above_high_watermark,
+                  std::function<void()> above_overflow_watermark)
```

Comment:
I think I'ld like to change the semantic of high watermark and overflow watermark so they fire when we hit the configured limit instead of exceeding it.  Can we name this on_overflow_watermark instead?

Reply:
I think that the semantics should be consistent between all three watermarks, whether it's on or below/above.
Does it make sense to have this first changed to LW/HW, and only then add the OW?

Reply:
#9499 already made the low-watermark callback name incorrect, sorry.

I don't know if changing high-watermark to trigger when equal to the limit makes sense.  I need to do some more research and discuss with others since the current behavior comes from a design that has some performance goals behind it.

I don't want to block you on naming issues.  It is possible that above makes sense for overflow watermark but not for high-watermark.

Reply:
At the moment we are using an overflow watermark multiplier (not setting the threshold directly).
In case the multiplier is set to 1, then the overflow threshold is equal to the high watermark threshold, and I think it makes sense to keep this strictly greater than, to avoid a case where the overflow triggered but not the high watermark.

That said, we can change both of OW and HW to trigger on equal.